pkg/streamingpromql/operators/aggregation.go

```diff
@@ -154,64 +153,72 @@ func (a *Aggregation) NextSeries(ctx context.Context) (types.InstantVectorSeries
+	thisGroup.floatSums = nil
+	thisGroup.floatPresent = nil
+	thisGroup.histogramSums = nil
+	thisGroup.histogramPointCount = 0
 
-			thisSeriesGroup.floatPresent, err = a.Pool.GetBoolSlice(steps)
-			if err != nil {
-				return types.InstantVectorSeriesData{}, err
-			}
-			thisSeriesGroup.floatSums = thisSeriesGroup.floatSums[:steps]
-			thisSeriesGroup.floatPresent = thisSeriesGroup.floatPresent[:steps]
+	groupPool.Put(thisGroup)
+	return seriesData, nil
+}
+
+func (a *Aggregation) constructSeriesData(thisGroup *group, start int64, interval int64) (types.InstantVectorSeriesData, error) {
+	floatPointCount := reconcilePointCount(thisGroup)
+	var floatPoints []promql.FPoint
+	var err error
+	if floatPointCount > 0 {
+		floatPoints, err = a.Pool.GetFPointSlice(floatPointCount)
+		if err != nil {
+			return types.InstantVectorSeriesData{}, err
 		}
 
-		if len(s.Histograms) > 0 && thisSeriesGroup.histogramSums == nil {
-			// First series for this group, populate it.
-			thisSeriesGroup.histogramSums, err = a.Pool.GetHistogramPointerSlice(steps)
-			if err != nil {
-				return types.InstantVectorSeriesData{}, err
+		for i, havePoint := range thisGroup.floatPresent {
+			if havePoint {
+				t := start + int64(i)*interval
+				floatPoints = append(floatPoints, promql.FPoint{T: t, F: thisGroup.floatSums[i]})
 			}
-			thisSeriesGroup.histogramSums = thisSeriesGroup.histogramSums[:steps]
 		}
+	}
 
-		for _, p := range s.Floats {
-			idx := (p.T - start) / interval
-			thisSeriesGroup.floatSums[idx] += p.F
-			thisSeriesGroup.floatPresent[idx] = true
+	var histogramPoints []promql.HPoint
+	if thisGroup.histogramPointCount > 0 {
+		histogramPoints, err = a.Pool.GetHPointSlice(thisGroup.histogramPointCount)
+		if err != nil {
+			return types.InstantVectorSeriesData{}, err
 		}
 
-		for _, p := range s.Histograms {
-			idx := (p.T - start) / interval
-			if thisSeriesGroup.histogramSums[idx] == nil {
-				// We copy here because we modify the histogram through Add later on.
-				// It is necessary to preserve the original Histogram in case of any range-queries using lookback.
-				thisSeriesGroup.histogramSums[idx] = p.H.Copy()
-				// We already have to do the check if the histogram exists at this idx,
-				// so we can count the histogram points present at this point instead
-				// of needing to loop again later like we do for floats.
-				thisSeriesGroup.histogramPointCount++
-			} else {
-				thisSeriesGroup.histogramSums[idx] = thisSeriesGroup.histogramSums[idx].Add(p.H)
+		for i, h := range thisGroup.histogramSums {
+			if h != nil {
+				t := start + int64(i)*interval
+				histogramPoints = append(histogramPoints, promql.HPoint{T: t, H: thisGroup.histogramSums[i]})
 			}
 		}
-
-		a.Pool.PutInstantVectorSeriesData(s)
-		thisSeriesGroup.remainingSeriesCount--
 	}
+	return types.InstantVectorSeriesData{Floats: floatPoints, Histograms: histogramPoints}, nil
+}
 
-	// Construct the group and return it
+func reconcilePointCount(thisGroup *group) int {
```

Comment:
I don't love this name, simply because it hides the fact that we also drop samples where we have both floats and histograms - any ideas for other names? Another option would be splitting it in two, one part for dealing with conflicts and the other to count the number of float points we have

Reply:
Yeah, that's a fair point. I've renamed it to `reconcileAndCountFloatPoints` and added a docstring to help explain. If it's too convoluted we can split out the functions.